interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,398 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const customDeviceSummary = `provides access to custom devices specified via the gadget snap`
+
+const customDeviceBaseDeclarationSlots = `
+  custom-device:
+    allow-installation: false
+    allow-connection:
+      plug-attributes:
+        content: $SLOT(custom-device)
+    deny-auto-connection: true
+`
+
+var (
+	// A cryptic, uninformative error message that we use only on impossible code paths
+	customDeviceInternalError = errors.New(`custom-device interface internal error`)
+
+	// Validating regexp for filesystem paths
+	customDevicePathRegexp = regexp.MustCompile(`^/[^"@]*$`)
```

Comment:
Should this exclude whitespace in the path?

Reply:
I think it's okay to allow whitespace since we quote everything